File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1)
        current = dummy
        carry = 0

        while l1 or l2 or carry:
            current_sum = 0
            
            if l1:
                current_sum += l1.val
                l1 = l1.next
            
            if l2:
                current_sum += l2.val
                l2 = l2.next
            
            if carry:
                current_sum += carry
            
            current.next = ListNode(current_sum % 10)
            current = current.next

            carry = current_sum // 10

        return dummy.next
            


        

File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let c = 0;
    let head = new ListNode();
    let current = head;

    while(l1 || l2 || c) {
        let sum = 0;
        if( l1 )
            sum += l1.val;

        if( l2 )
            sum += l2.val;
            
        if( c )
            sum += c;


        if( sum > 9 ){
            let remainder = sum - 10;
            c = 1;
            let newNode =  new ListNode(remainder);
            current.next = newNode;
            current = newNode            
        } else {
            c = 0;
            let newNode =  new ListNode(sum);
            current.next = newNode;
            current = newNode
        }

        l1 = l1 ? l1.next : l1;
        l2 = l2 ? l2.next : l2;
    }

    return head.next;
};